### ISSUE TITLE
not reloading config on consul_catalog health check changes

### ISSUE THREAD

**Issue by tyler-johnson** [containous/traefik#1107]
### What version of Traefik are you using (`traefik version`)?

v1.1.2

### What is your environment & configuration (arguments, toml...)?

Traefik is running as a normal process in Ubuntu 16.10. It is connected to a local Consul client agent which is connected to a Consul/Nomad cluster.

```bash
traefik -c /etc/traefik/traefik.toml
```

```ini
logLevel = "INFO"

[web]
address = ":8080"

[consul]
endpoint = "127.0.0.1:8500"
watch = true
prefix = "traefik"

[consulCatalog]
endpoint = "127.0.0.1:8500"
domain = "consul.localhost"
prefix = "traefik"
```

### What did you do?

I pushed a Nomad job with some Consul tags for Traefik. The Nomad job has a health check that is initially failing, but passes once the service is fully online. Consul registers the service correctly, with the Traefik tag and passing health checks eventually.

### What did you expect to see?

Traefik to update when the health check passes and attach some consul services to frontends.

### What did you see instead?

Sometimes it goes through, other times it does not. Traefik will still give a 404 and the web ui does not show my new service. I have also waited a decent amount of time and nothing happens. If I push a different Nomad job, Traefik will update with the missing service. When I totally disable health checks on a Consul service, Traefik will update properly.

**Comment by vholovko** [comment]
Hey,

I have investigated a bit the current situation.

Basically consul_catalog provider is relying on blocking queries to `/v1/catalog/services` consul endpoint.
https://www.consul.io/docs/agent/http/catalog.html#catalog_services
The query itself might look like this one:

GET /v1/catalog/services?index=108026&wait=15000ms

where traefik config change triggers each time whenever `X-Consul-Index` is changed.

A `tcpdump` within consul container looks like this
```
# tcpdump -A -l -i eth2 | grep "?index="
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth2, link-type EN10MB (Ethernet), capture size 262144 bytes
.......qGET /v1/catalog/services?index=108026&wait=15000ms HTTP/1.1
........GET /v1/catalog/services?index=108026&wait=15000ms HTTP/1.1
GET /v1/catalog/services?index=108033&wait=15000ms HTTP/1.1
..~ ..~ GET /v1/catalog/services?index=108033&wait=15000ms HTTP/1.1
```

If you test it with the simple GET request from Postman you will find out that it is reacting on service registration/deregistration events.
Seems fine and fair enough but turns out that '/v1/catalog/services' endpoint DO NOT REACT on service health change!
Like in a case where health state is changing from warning/critical to passing.

The closest plausible query that is more sensible enough is `/v1/health/state/passing`.
https://www.consul.io/docs/agent/http/health.html#health_state
It can detect a change in amount of healthy services and I think this is perfectly suitable for our purposes.

A `tcpdump` within consul container should look like this
```
# tcpdump -A -l -i eth2 | grep "?index="
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth2, link-type EN10MB (Ethernet), capture size 262144 bytes
..m{..m{GET /v1/health/state/passing?index=108034&wait=15000ms HTTP/1.1
..s...s.GET /v1/health/state/passing?index=108034&wait=15000ms HTTP/1.1
..x...x.GET /v1/health/state/passing?index=108128&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
...&...&GET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
...C...CGET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
...x...xGET /v1/health/state/passing?index=108137&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108149&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108161&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108161&wait=15000ms HTTP/1.1
........GET /v1/health/state/passing?index=108218&wait=15000ms HTTP/1.1
```

This is all my finding for now. 
I can provide the pull request but I don't very familiar with the GO language so my solution might not be the optimal one :)

**Comment by emilevauge** [comment]
@tyler-johnson A fix has been done in PR #1241 by @vholovko. Do you have the possibility to test it before we merge it ?

**Comment by tyler-johnson** [comment]
Unfortunately I do not as I went back to plain nginx and docker. I had lots of other issues with Nomad so I abandoned it, but I still really want to use Traefik so within the next few weeks I will be attempting a full integration of my app with Consul and Traefik.

Thanks for the work on this!

**Comment by jzvelc** [comment]
I just tested this and it works! Thank you! I hope it gets merged soon...

**Comment by maticmeznar** [comment]
Has there been any progress on this?

**Comment by hamann** [comment]
I just seen a similiar behaviour. Scheduled a new nomad job with two instances, both are healthy in consul, but just one backend is picked up by traefik. Could that be related?

**Comment by vholovko** [comment]
@hamann

Yes, it is related.
 
Services by default in consul right after registration appears in unhealthy state (latency for health state first resolution).
In that same moment consul_provider is able to detect service changes, but it does not pick up unhealthy ones.
Then after service becomes healthy nothing happens.

Actually, I was struggling with the same situation before looking more closely into the issue.

**Comment by pierredavidbelanger** [comment]
I hope this issue is fixed soon!
I really like the simplicity of Traefik and Consul.

I have a workaround, it is rather obvious, but I will share it anyway.

For now my workaround is to register the service (`/v1/agent/service/register`) **without** check (instead of registering the service and its check in the same call), than immediately after the service registration is done, I register a check (`/v1/agent/check/register`) for this service. I use the `DeregisterCriticalServiceAfter` option on the check to make sure Consul deregister the service when the check fail (so Traefik is able reload its config).

I guess this is working because as soon as the new service is registered without check, the long polling returns with the new service, Traefik is able to see it as healty, so it reload its config.

There is probably a race condition somewhere here that I am not aware of, but for now it is working for me.

**Comment by hamann** [comment]
I tested it with 1.3-rc1 and now service instances where added as expected. But there's quite a delay (about 20 seconds) between when an instance appears in consul as healthy and traefik decides to add it (if checking web ui is reliable enough for measuring).

**Comment by vholovko** [comment]
@hamann

Thank you for testing,

This extra query for changes was made as an auxiliary call to /health endpoint after /services endpoint long polling. 
In worst case, long polling takes full 15 seconds + 5 second traefik configuration taking action after health change detected.

It is possible to make an improvement here to do a long polling for services and for the health in the separate coroutines and react on either trigger independently by using same channel for both coroutines.

**Comment by hamann** [comment]
@vholovko Thanks! The 15 seconds come from [here](https://github.com/containous/traefik/blob/master/provider/consul/consul_catalog.go#L24)? If yes, any concerns about changing that to 1 second in a custom build for my purposes until the improvements where applied?

**Comment by vholovko** [comment]
@hamann yes, it comes from there. IMHO, don't see any concerns about it and I believe you can do so to reduce the overall timing...

**Comment by rokka-n** [comment]
Shouldn't be a problem since application instance will be talking to local agent anyways.

**Comment by hamann** [comment]
fabio goes another route - instead of watching services endpoint, it watches `/v1/health/state/any` and then filters for relevant services. 
https://github.com/fabiolb/fabio/blob/master/registry/consul/service.go#L22
Sounds more reasonable, doesn't it?

**Comment by vholovko** [comment]
@hamann
sounds reasonable, 
but what about services that are registering without any health checks ?
can we detect them with /v1/health/state/any ?

**Comment by pierredavidbelanger** [comment]
@vholovko 

https://www.consul.io/api/health.html#list-checks-in-state

> This endpoint returns the checks in the state provided on the path.

So I guess not.

**Comment by vholovko** [comment]
thanks @pierredavidbelanger, 

/v1/health/state/any is not a silver bullet in all possible service registrations, and we should consider these duality by listening both in the same time: `/v1/health/state/any` and `/catalog/services`. 

Next question: do we only interested in healthy services (the passing ones) ? `warning` services can also we workable enough to consider them in `traefik` config.

Turns out that `consul_catalog` should be changed to not filter only `passing` ones but might be only filter out criticals.

**Comment by hamann** [comment]
@vholovko Who has services without health checks?

**Comment by hamann** [comment]
@vholovko I don't know of one case, where `warning` makes sense (didn't even know that this exists in terms of consul health check). I guess filtering for `passing` is completly adequate.

**Comment by emilevauge** [comment]
@vholovko @hamann @pierredavidbelanger should we leave this issue open? Is there a fix needed in Traefik?

**Comment by pierredavidbelanger** [comment]
@emilevauge 

The issue is still there in Traefik (at least with 1.2).

I found a workaround for my services that manually register themself (see [this comment](https://github.com/containous/traefik/issues/1107#issuecomment-298813175)) 

Also, I found a fix on the [service registration side](https://github.com/gliderlabs/registrator/blob/master/docs/user/backends.md#consul-initial-health-check-status)

So, I am ok if the fix does not come from Traefik.

**Comment by michaelpjohnson** [comment]
@emilevauge Would this affect the fact that I don't see "consul" and "consul_catalog" tabs on my providers page? I can see the frontends/backends via a `curl` to the `/api` endpoint, but not via the UI itself. Not a showstopper, but makes the UI at the moment.

**Comment by ldez** [comment]
@pierredavidbelanger could you try with v1.3.0-rc3?

**Comment by pierredavidbelanger** [comment]
@ldez indeed, `traefik:1.3.0-rc3-alpine` is now working without my workarounds. It seems it is now able to pick up services as soon as health check is passing, is that correct ?

**Comment by ldez** [comment]
yes, see https://github.com/containous/traefik/issues/1107#issuecomment-300492280 for more information.

**Comment by ldez** [comment]
Closed by #1241

**Comment by vholovko** [comment]
@hamann  @ldez @emilevauge 
Sorry for always being late.
I have done the improvement I was talking about earlier.
Thus this thread is already closed we can just jump into #1514 which is more relevant to proposed pull request #1694
